private/bufpkg/bufplugin/bufplugindocker/docker_test.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker_test
+
+import (
+	"context"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginconfig"
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufplugindocker"
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginref"
+	"github.com/docker/docker/client"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+var dockerEnabled = false
+
+const examplePluginIdentity = "plugins.buf.build/library/go"
+
+func Test_BuildSuccess(t *testing.T) {
+	t.Parallel()
+	if !dockerEnabled {
+		t.Skip("docker disabled")
+	}
+	if testing.Short() {
+		t.Skip("docker tests disabled in short mode")
+	}
+	dockerClient := createClient(t)
+	response, err := buildDockerPlugin(t, dockerClient, "testdata/success/Dockerfile")
+	require.Nil(t, err)
+	assert.Truef(
+		t,
+		strings.HasPrefix(response.Image, examplePluginIdentity+":"),
+		"image name should begin with: %q, found: %q",
+		examplePluginIdentity,
+		response.Image,
+	)
+	assert.NotEmpty(t, response.Digest)
+}
+
+func Test_BuildFailure(t *testing.T) {
+	t.Parallel()
+	if !dockerEnabled {
+		t.Skip("docker disabled")
+	}
+	if testing.Short() {
+		t.Skip("docker tests disabled in short mode")
+	}
+	dockerClient := createClient(t)
+	_, err := buildDockerPlugin(t, dockerClient, "testdata/success/Dockerfile")
+	require.Nil(t, err)
+}
+
+func TestMain(m *testing.M) {
+	if cli, err := client.NewClientWithOpts(client.FromEnv); err == nil {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+		if _, err := cli.Ping(ctx); err == nil {
+			dockerEnabled = true
```

Comment:
I wonder if we should require that these tests are run in the CI environments we care about though. Otherwise if there's ever a runner that can't successfully ping, we'll have a bunch of silent failures and the build will be green because the tests will have been skipped.

Reply:
Since Docker isn't documented as a requirement to run tests, I was reluctant to make these required for others building/testing Buf. I do think we'd want to catch in CI though if these were for some reason disabled (i.e. we switched to a different CI system that didn't provide Docker).

What do you think about having an environment variable `TEST_DOCKER_FORCE=true` which could be set in our Github actions CI workflows to ensure we run these tests each time? If that was set, we'd skip the ping check and just enable these tests.

Reply:
An environment variable would work - we can handle this in a follow-up though. As long as this doesn't yield flaky results in our CI right now, I think it's fine to mark it with a `TODO` and revisit. If this is useful for testing locally in the interim, that's fine with me.